llvm-passes/ObfuscateVar/ObfuscateVar.cpp

```diff
@@ -0,0 +1,210 @@
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IRBuilder.h"
+
+#ifndef NDEBUG
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/Debug.h"
+#endif
+
+#include <map>
+#include <utility> //std::pair, std::make_pair
+#include <random>
+
+
+using namespace llvm;
+
+namespace {
+  typedef uint32_t typeInter;
+  typedef std::pair<AllocaInst*,AllocaInst*> typeMapValue;
+  typedef std::map<Value*,typeMapValue> typeMap;
+  
+class ObfuscateVar : public BasicBlockPass {
+public:
+  
+  static char ID;
+  ObfuscateVar() : BasicBlockPass(ID) {}
+
+  virtual bool runOnBasicBlock(BasicBlock &BB) override {
+    /* 
+       Principe de l'obfuscation (d'après https://www.cs.ox.ac.uk/files/2936/RR-10-02.pdf):
+       Etant donné une variable X, on l'a "split" en deux variables selon
+       la formule :
+       - x1 = X%10
+       - x2 = X//10
+       
+       X=X+a sera donc transformé en :
+       - x2 = (10*x2+x1+a)//10
+       - x1 = x1+a mod 10
+
+       Pour "reconstruire" la variable : 
+       X = 10*x2+x1
+
+    */
+    
+    for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),end = BB.end(); I != end; ++I) {
+      
+      Instruction &Inst = *I;
+      if(BinaryOperator *Binop = dyn_cast<BinaryOperator>(&Inst)) {
+        
+        if(Binop->getOpcode() == Instruction::Add){
+          
+          if (isa<Constant>(Binop->getOperand(1))) {// add X,#CONSTANT
+            
+            if(Value* v = replaceAddOp(Inst)){
+              Inst.replaceAllUsesWith(v);
+            }
+
+          }//TODO : add X,Y for all X,Y
+        }//TODO : sub,div,mul...
+          
+         
+      }else if(StoreInst *sreInst = dyn_cast<StoreInst>(&Inst)){
+        if(!isa<PointerType>(sreInst->getValueOperand()->getType())){
+          dbgs() << "Store Instruction: "<<*sreInst << "\n";
+          registerVar(*sreInst);
+        }
+        
+      }
+    }
+
+    return true;
+  }
+
+private:
+
+  // Enregistre dans une map les x1,x2 associés a une variable
+  // TODO : return value* and replace
+  void registerVar(StoreInst &S){
+    bool isVolatile = false;
+    dbgs() << "operand value = " << *S.getValueOperand() << "\n";
+      
+    BasicBlock *BB = S.getParent();
+    Type *IntermediaryType = IntegerType::get(BB->getContext(),sizeof(typeInter)*8);
+    Value *varPointer = S.getPointerOperand();
+      
+    //dbgs() << "Store Inst :\n\t" << S << "\n\tValue is : " << *C0 <<  "\n\Address is : " << *varPointer << "\n\n";
+
+    AllocaInst* alloX1 = new AllocaInst(IntermediaryType, 0,"x1");
+    AllocaInst* alloX2 = new AllocaInst(IntermediaryType, 0,"x2");
+
+    BB->getInstList().insert(S, alloX1); //%x1 = alloca i32, align 4
+    BB->getInstList().insert(S, alloX2); //%x2 = alloca i32, align 4
```

Comment:
``AllocaInst`` has a constructor that inserts in the instruction directly, but you'll be better off using ``IRBuilder<>``